File: 4-player-AI-written-engine/scripts/simulate.py
"""Random legal demo / benchmark. No game rules live here."""
import argparse
import json
import time
from hokm import Engine, Phase

class Driver:
    def __init__(self):
        self.x = 0x12345678
    def next(self):
        x = self.x
        x ^= (x << 13) & 0xffffffff
        x ^= x >> 17
        x ^= (x << 5) & 0xffffffff
        self.x = x
        return x
    def index(self, n):
        threshold = ((-n) & 0xffffffff) % n
        while True:
            r = self.next()
            if r >= threshold:
                return r % n

def run(matches=1000, seed=123, mode="plain"):
    driver = Driver()
    steps = rounds = checksum = 0
    trace = []
    start = time.perf_counter()
    for i in range(matches):
        e = Engine(seed + i)
        while e.phase != Phase.MATCH_OVER:
            if e.phase == Phase.ROUND_OVER:
                e.start_next_round()
                continue
            if mode == "observe":
                o = e.observe(e.current_actor)
                checksum += len(o.hand) + len(o.history)
            actions = e.legal_actions()
            action = actions[driver.index(len(actions))]
            e.apply_id(action)
            steps += 1
            if mode == "trace":
                trace.append(action)
            if e.phase in (Phase.ROUND_OVER, Phase.MATCH_OVER):
                rounds += 1
                checksum += e.round_result.points
                if mode == "demo":
                    print(f"Round {rounds}: team {e.round_result.winner}, points {e.round_result.points}")
    seconds = time.perf_counter() - start
    result = dict(mode=mode, seed=seed, matches=matches, steps=steps, rounds=rounds,
                  seconds=seconds, steps_per_second=steps/seconds,
                  rounds_per_second=rounds/seconds, checksum=checksum)
    return result, trace

if __name__ == "__main__":
    p = argparse.ArgumentParser()
    p.add_argument("--matches", type=int, default=1000)
    p.add_argument("--seed", type=int, default=123)
    p.add_argument("--mode", choices=["plain", "observe", "trace", "demo"], default="plain")
    args = p.parse_args()
    if args.matches <= 0:
        p.error("matches must be positive")
    result, trace = run(args.matches, args.seed, args.mode)
    if args.mode == "trace":
        print(*trace)
    print(json.dumps(result))

File: 4-player-AI-written-engine/scripts/simulate.cpp
#include "hokm/engine.hpp"
#include <chrono>
#include <iostream>
#include <string>
using namespace hokm;
// Same small test-driver PRNG in C++/Python: identical workload, separate from engine.
struct Driver {
    std::uint32_t x;
    std::uint32_t next() {
        x ^= x << 13;
        x ^= x >> 17;
        x ^= x << 5;
        return x;
    }
    int index(int n) {
        std::uint32_t r, threshold = std::uint32_t(-std::uint32_t(n)) % n;
        do {
            r = next();
        } while (r < threshold);
        return r % n;
    }
};
int main(int argc, char **argv) {
    try {
        int matches = argc > 1 ? std::stoi(argv[1]) : 1000;
        std::uint64_t seed = argc > 2 ? std::stoull(argv[2]) : 123;
        std::string mode = argc > 3 ? argv[3] : "plain";
        if (matches <= 0 ||
            (mode != "plain" && mode != "observe" && mode != "trace" && mode != "demo"))
            throw std::invalid_argument("usage: hokm_sim MATCHES SEED [plain|observe|trace|demo]");
        bool obs = mode == "observe", trace = mode == "trace", demo = mode == "demo";
        Driver driver{0x12345678};
        std::uint64_t steps = 0, rounds = 0, checksum = 0;
        auto start = std::chrono::steady_clock::now();
        for (int i = 0; i < matches; ++i) {
            Engine e(seed + i);
            while (e.phase() != Phase::MatchOver) {
                if (e.phase() == Phase::RoundOver) {
                    e.start_next_round();
                    continue;
                }
                if (obs) {
                    auto o = e.observe(e.current_actor());
                    checksum += o.hand.size() + o.history.size();
                }
                auto a = e.legal_actions();
                int id = a[driver.index(static_cast<int>(a.size()))];
                e.apply_id(id);
                ++steps;
                if (trace)
                    std::cout << id << ' ';
                if (e.phase() == Phase::RoundOver || e.phase() == Phase::MatchOver) {
                    ++rounds;
                    auto r = e.round_result();
                    checksum += r.points;
                    if (demo)
                        std::cout << "Round " << rounds << ": team " << r.winner << ", points "
                                  << r.points << '\n';
                }
            }
        }
        double seconds =
            std::chrono::duration<double>(std::chrono::steady_clock::now() - start).count();
        if (trace)
            std::cout << '\n';
        std::cout << "{\"mode\":\"" << mode << "\",\"seed\":" << seed << ",\"matches\":" << matches
                  << ",\"steps\":" << steps << ",\"rounds\":" << rounds
                  << ",\"seconds\":" << seconds << ",\"steps_per_second\":" << steps / seconds
                  << ",\"rounds_per_second\":" << rounds / seconds << ",\"checksum\":" << checksum
                  << "}\n";
    } catch (const std::exception &e) {
        std::cerr << e.what() << '\n';
        return 1;
    }
}
